### PR TITLE
Issue 6600: LTS - Ability to simulate slow LTS

### LINKED ISSUE #6600: LTS - Ability to simulate slow LTS
**Is your feature request related to a problem? Please describe.**
We need to introduce delays for testing slow LTS.

**Describe the solution you'd like**

- Add new config settings to StorageExtra to enable slow mode and take delay as duration
- StorageLoader should create ChunkedSegmentStorageConfig (similar to line 44)
- Depending on whether slow mode is true, the StorageLoader::load returns wrapped instance. 
- Implement SlowStorage class implements Storage interface and forwards the calls to wrapped storage, However it introduces delays and timeouts as specified in config/parameters above.  Use Futures::delayedFuture passing executor and delay. Use CompletableFuture::composeAsync appropriately. 
- Implement SlowChunkStorage class implements ChunkStorage interface and forwards the calls to wrapped ChunkStorage, However it introduces delays and timeouts as specified in config/parameters above.  Use Futures::delayedFuture passing executor and delay. Use CompletableFuture::composeAsync appropriately. 

**Describe alternatives you've considered**
Adhoc code changes to test slowness are not desirable.

**Additional context**

### PR DESCRIPTION
**Pull request by AJadhav29**
**Change log description**  
LTS -  Introduce delays and time out for testing.

**Purpose of the change**  
Fixes #6600

**What the code does**  
The code adds new config settings to ChunkedSegmentStorageConfig 
- inject.delay.enbale type boolean
- inject.delay.mean type int
- inject.delay.stddev type int
- inject.delay.cycletime type int (time taken by sinusoidal wave to complete one cycle)
- inject.delay.distribution.type type string (depending upon type of delay generated, the type of distribution can be 'Normal', 'Fixed' or 'Sinusoidal' )
- StorageLoader should create ChunkedSegmentStorageConfig (similar to line 44)
- Depending on whether inject.delay.enbale is true and inject.storage is true, the StorageLoader::load returns wrapped instance.
- Add following config values to make storage slow for testing:
```
storageextra.slow.enable = true
storageextra.slow.latency.mean.ms = 10
storageextra.slow.latency.std.ms = 10
```

Implements FlakyStorageWrapper class implements Storage interface and forwards the calls to wrapped storage,
- However it introduces delays and timeouts as specified in config/parameters above
- FlakyStorageWrapper constructor takes 2 parameters int meanDelay , int stdDeviation
- Don't use Thread.sleep , instead use Futures::delayedFuture passing excecutor and delay. Use CompletableFuture::composeAsync appropriately. 
- Generate delay value using = mean + Random::nextGaussian() * stddev
- Generate a fixed delay which is configured as always the same delay. 
- Generate a sinusoidal delay which induces delays from 0 up to storageextra.slow.sinusoidal.latency.ms increasingly up to a period of time storageextra.slow.sinusoidal.period.ms, and then goes progressively down to 0 again. This can be easily generated by using Math.sin() function based on the time period defined, so we generate "waves" of lower and higher latency of LTS. The time calculated as  = (current time - start time) % cycle time (where cycle time is time taken by sinusoidal wave to complete one cycle.) And the delay generated calculated as = (1 + Math.sin(Math.toRadians(timeCal))) * mean

**How to verify it**  
All the tests should pass

### CODE REVIEW

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowStorage.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * Mock for StorageFactory. Contents is destroyed when object is garbage collected.
+ */
+public class SlowStorage implements Storage {
+    @Getter
+    final Storage inner;
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowStorage.java:36]: make  `final protected` for these 3 fields

**Review by sachin-j-joshi — CHANGES_REQUESTED** [review]
I have requested a few minor changes.
Looks good otherwise

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

**Comment by sachin-j-joshi** [comment]
@AJadhav29 Please update the PR description to reflect current design/changes.

**Review by RaulGracia — CHANGES_REQUESTED** [review]
Thanks, this is useful for our tests from now on. However, I liked more an approach similar to the original description of the PR. The current approach is too simplistic and this limits our testing capabilities:  that is, having a fixed delay will limit us to see whether the system can recover automatically when LTS becomes working properly again (with the current approach, we would need to change the configuration of the fixed delay to 0 and restart all instances).

I suggest the following:
- Add another configuration property that is something similar to `storageextra.slow.delaytype`, so we can use different types of delays (or delay generators).
- At the moment, we could have 2 simple ones: "fixed", which is the one implemented and always generates the same delay (it can be configured via something like `storageextra.slow.fixeddelay.latency.ms`). Then, a "sinusoidal" delay, which induces delays from 0 up to `storageextra.slow.sinusoidal.latency.ms` increasingly up to a period of time `storageextra.slow.sinusoidal.period.ms`, and then goes progressively down to 0 again. This can be easily generated by using `Math.sin()` function based on the time period defined, so we generate "waves" of lower and higher latency of LTS.

The previous suggestion is important because of 3 reasons: i) We need a way to really test that PR https://github.com/pravega/pravega/pull/6597 works, and one of the main aspects to tests is not just it can throttle when it is needed, but also that it can stop throttling when LTS works well again. ii) In general, we need to test that Pravega is self-healing after some period of LTS being unavailable. This "sinusoidal" delay generator is way to achieve this automatically, for instance during our longevity runs. iii) This can help us to measure the max time Pravega can tolerate LTS being slow (by progressively increasing `storageextra.slow.fixeddelay.latency.ms`), if there is actually a max time for that.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java [hunk 2]
```diff
@@ -30,8 +30,15 @@
 
     public static final Property<Boolean> STORAGE_NO_OP_MODE = Property.named("noOp.mode.enable", false, "storageNoOpMode");
     public static final Property<Integer> STORAGE_WRITE_NO_OP_LATENCY = Property.named("noOp.write.latency.milliseconds", 20, "storageWriteNoOpLatencyMillis");
+
+    public static final Property<Boolean> STORAGE_SLOW_MODE = Property.named("slow.enable", false);
+    public static final Property<Integer> STORAGE_SLOW_MODE_LATENCY_MEAN = Property.named("slow.latency.mean.ms", 500);
+    public static final Property<Integer> STORAGE_SLOW_MODE_LATENCY_STD_DEV = Property.named("slow.latency.std.ms", 500);
+    public static final Property<String> STORAGE_SLOW_MODE_DISTRIBUTION_TYPE = Property.named("slow.type", "Normal");
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java:37]: Please add property public static final Property<Integer> STORAGE_SLOW_MODE_LATENCY_CYCLE_TIME = Property.named("slow.latency.cycle.ms", 300000); // 5 minutes

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java [hunk 3]
```diff
@@ -44,6 +51,24 @@
...
+    private final boolean slowModeEnabled;
+
+    /**
+     * Latency in milliseconds applied for slow storage mode.
+     */
+    @Getter
+    private final int slowModeLatencyMeanMillis;
+
+    @Getter
+    private final int slowModeLatencyStdDevMillis;
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java:67]: please add 
```
@Getter
    private final int slowModeLatencyCycleTimeMillis;
```

and initialize it in constructor

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+public class SlowDelaySuppliers {
+    static Supplier<Duration> getDurationSupplier(StorageExtraConfig config) {
+        if (config.getDistributionType().equals("Fixed")) {
+            return new FixedDelaySupplier(config.getSlowModeLatencyMeanMillis());
+        }
+        if (config.getDistributionType().equals("Normal")) {
+            return new GaussianDelaySupplier(config.getSlowModeLatencyMeanMillis(), config.getSlowModeLatencyStdDevMillis());
+        }
+        if (config.getDistributionType().equals("Sinusoidal")) {
+            return new SinusoidalDelaySupplier(config.getSlowModeLatencyMeanMillis(), 10);
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java:38]: please use `config.getSlowModeLatencyCycleTimeMillis()` here

**Review by sachin-j-joshi — CHANGES_REQUESTED** [review]

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.time.Duration;
+import java.util.Random;
+import java.util.function.Supplier;
+
+/**
+ * States different types of delays generated in the working of LTS.
+ */
+public class SlowDelaySuppliers {
+    static Supplier<Duration> getDurationSupplier(StorageExtraConfig config) {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java:30]: Missing comment

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    static class GaussianDelaySupplier implements Supplier<Duration> {
+        private final long mean;
+        private final long stdDev;
+        private final Random random = new Random();
+        @Override
+        public Duration get() {
+            return Duration.ofMillis(calculateValue(() -> random.nextGaussian()));
+        }
+
+        public long calculateValue(Supplier<Double> randomSupplier) {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java:56]: may be public is not needed here.

Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java:56]: missing comment

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            this.cycleTime = cycleTime;
+        }
+
+        @Override
+        public Duration get() {
+            final long returnValue = calculateValue(System.currentTimeMillis(), startTime, cycleTime);
+            return Duration.ofMillis(returnValue);
+        }
+
+        long calculateValue(long currentTime, long startTime, long cycleTime) {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java:81]: missing comment

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/mocks/SlowStorageTests.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        innerStorage.getGarbageCollector().initialize(new InMemoryTaskQueueManager()).join();
+        return new SlowStorage(innerStorage, executor, StorageExtraConfig.builder()
+                .with(StorageExtraConfig.STORAGE_SLOW_MODE, true)
+                .with(StorageExtraConfig.STORAGE_SLOW_MODE_DISTRIBUTION_TYPE, "Normal")
+                .with(StorageExtraConfig.STORAGE_SLOW_MODE_LATENCY_STD_DEV, 0)
+                .with(StorageExtraConfig.STORAGE_SLOW_MODE_LATENCY_MEAN, 0)
+                .build());
+    }
+
+    public static class SlowRollingTests extends RollingStorageTests {
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/mocks/SlowStorageTests.java:58]: missing comment

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/noop/StorageExtraConfigTest.java [hunk 9]
```diff
@@ -43,4 +43,18 @@ public void testNoOpSwitch() {
...
+    public void testSlowModeEnabled() {
+        ConfigBuilder<StorageExtraConfig> builder = StorageExtraConfig.builder();
+        builder.with(Property.named("slow.enable"), true);
+        assertEquals(true, builder.build().isSlowModeEnabled());
+    }
+
+    @Test
+    public void testSlowModeLatency() {
+        ConfigBuilder<StorageExtraConfig> builder = StorageExtraConfig.builder();
+        builder.with(Property.named("slow.latency.ms"), 500);
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/noop/StorageExtraConfigTest.java:57]: I don't think the property name here correct.
It should be `storageextra.slow.latency.mean.ms`
also please value other than default like 42

**Review by sachin-j-joshi — CHANGES_REQUESTED** [review]

File: segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/noop/StorageExtraConfigTest.java [hunk 10]
```diff
@@ -54,7 +54,7 @@ public void testSlowModeEnabled() {
     @Test
     public void testSlowModeLatency() {
         ConfigBuilder<StorageExtraConfig> builder = StorageExtraConfig.builder();
-        builder.with(Property.named("slow.latency.ms"), 500);
-        assertEquals(500, builder.build().getSlowModeLatencyMeanMillis());
+        builder.with(Property.named("storageextra.slow.latency.mean.ms"), 42);
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/noop/StorageExtraConfigTest.java:57]: `slow.latency.mean.ms`

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java [hunk 11]
```diff
@@ -32,9 +32,10 @@
     public static final Property<Integer> STORAGE_WRITE_NO_OP_LATENCY = Property.named("noOp.write.latency.milliseconds", 20, "storageWriteNoOpLatencyMillis");
 
     public static final Property<Boolean> STORAGE_SLOW_MODE = Property.named("slow.enable", false);
-    public static final Property<Integer> STORAGE_SLOW_MODE_LATENCY_MEAN = Property.named("slow.latency.mean.ms", 500);
+    public static final Property<Integer> STORAGE_SLOW_MODE_LATENCY_MEAN = Property.named("slow.latency.mean.ms", 42);
```
Comment (sachin-j-joshi) [inline segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/noop/StorageExtraConfig.java:35]: please keep it 500

**Review by sachin-j-joshi — APPROVED** [review]
LGTM

**Comment by AJadhav29** [comment]
> @sachin-j-joshi the change for adding what I mentioned is minimal and this can enable us to start testing directly. I think it is perfectly doable in this PR just to add the property to select the delay generator and add 2 (fixed and sinusoidal ones). For more complex delay generators, I agree that we can add them later, but let's not delay what we need now for doing proper tests of scenarios where LTS is not operating correctly.
 @RaulGracia I have tried making the changes by adding 2 delay generators- fixed and sinusoidal, as suggested. Please have a look at it and let me know if any further changes or modifications are required.

**Comment by AJadhav29** [comment]
@RaulGracia I have tried making the changes by adding 2 delay generators- fixed and sinusoidal, as suggested. Please have a look at it and let me know if any further changes or modifications are required.

**Review by RaulGracia — APPROVED** [review]
Just one minor comment, once you address it we can merge it.

File: segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+ * Configuration properties stated in {@link StorageExtraConfig} are used to inject different types of delays (or delay generators).
+ */
+public class SlowDelaySuppliers {
+    /**
+     * Factory method that creates {@link Supplier<Duration>} based on {@link StorageExtraConfig}.
+     * @param config config to use.
+     * @return instance of {@link Supplier<Duration>}.
+     */
+    static Supplier<Duration> getDurationSupplier(StorageExtraConfig config) {
+        if (config.getDistributionType().equals("Fixed")) {
```
[the remark this answers is not in the document]
Comment (AJadhav29) [reply segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/mocks/SlowDelaySuppliers.java]: @sachin-j-joshi @RaulGracia I have made the changes as suggested by using enumerates for the type of distribution for LTS. It looks good now to me. Please review it and let me know if any further changes are required.

**Comment by AJadhav29** [comment]
@sachin-j-joshi @RaulGracia Please have a look at the changes made. I have added enumerate for distribution type of the delay for LTS. Let me know if any further changes are required.

**Comment by sachin-j-joshi** [comment]
Nice work! 
Just a small change please rename StorageDistributionType to StorageDelayDistributionType.

**Comment by sachin-j-joshi** [comment]
This change is ready merge 


Except for unrelated flaky test #6626  has started failing constantly since last few days.

**Pull request merged**